power_domain: gpio: respect timing parameters

JordanYates opened the pull request:
Update the GPIO power domain to respect the on-off and startup times specified in devicetree.

Commit:
power_domain: gpio: remove unused strings

Remove unused power domain transition strings.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
power_domain: gpio: use on-off and startup time

Respect the configured values for how long the domain takes to turn on,
and how long the domain needs to be off for before it can be repowered.

As these actions can block, guard the transition function with
pm_device_action_can_block. To avoid system PM being able to turn the
domain off but not back on again, guard the entire implementation.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

ceolin commented on drivers/power_domain/power_domain_gpio.c:
```diff
@@ -22,28 +22,26 @@ struct pd_gpio_config {
 		gpio_pin_set_dt(&cfg->enable, 1);
 		LOG_DBG("%s is now ON", dev->name);
+		/* Wait for domain to come up */
+		k_sleep(K_USEC(cfg->startup_delay_us));
```
This action can be triggered from the idle thread, you can't sleep.

ceolin reviewed (changes requested):
suspend/resume devices can be triggered from the idle thread when the device is not using device runtime power management. In this context, we can't sleep.

JordanYates commented:
> suspend/resume devices can be triggered from the idle thread when the device is not using device runtime power management. In this context, we can't sleep.

This applies to external devices as well? If so this is going to be frankly unacceptable, it is impossible to write power transitions for most external drivers without blocking calls of some kind, whether that is `k_sleep`, `spi_transceive`, etc. cc @gmarull

ceolin commented:
> > suspend/resume devices can be triggered from the idle thread when the device is not using device runtime power management. In this context, we can't sleep.
> 
> This applies to external devices as well? If so this is going to be frankly unacceptable, it is impossible to write power transitions for most external drivers without blocking calls of some kind, whether that is `k_sleep`, `spi_transceive`, etc. cc @gmarull

If that is called from the idle thread yes ... you can't sleep there or block. This operation is if called through device runtime though. We need to figure it out indeed.

JordanYates commented:
> If that is called from the idle thread yes ... you can't sleep there or block. This operation is ok if called through device runtime though. We need to figure it out indeed but that is not a trivial problem to be fair.

Why is the idle thread trying to run transitions on external devices though?

ceolin commented:
> > If that is called from the idle thread yes ... you can't sleep there or block. This operation is ok if called through device runtime though. We need to figure it out indeed but that is not a trivial problem to be fair.
> 
> Why is the idle thread trying to run transitions on external devices though?

Can you be more specific about "external devices" ? From the kernel perspective what is the difference from these "external" drivers ? The PM subsytem will suspend / resume all devices that are not enabled for device runtime power management, not busy, not being used as wake up source and not have their state locked, depending on the power state.

ceolin commented:
> > > suspend/resume devices can be triggered from the idle thread when the device is not using device runtime power management. In this context, we can't sleep.
> > 
> > 
> > This applies to external devices as well? If so this is going to be frankly unacceptable, it is impossible to write power transitions for most external drivers without blocking calls of some kind, whether that is `k_sleep`, `spi_transceive`, etc. cc @gmarull
> 
> If that is called from the idle thread yes ... you can't sleep there or block. This operation is ok if called through device runtime though. We need to figure it out indeed but that is not a trivial problem to be fair.

One possibility is to iterate over the devices N times while there are devices busy and in each iteration re-check which action to take based on the remaining time, need to think a little bit more about it though.

JordanYates commented:
> Can you be more specific about "external devices" ? From the kernel perspective what is the difference for these "external" devices ? The PM subsytem will suspend / resume all devices that are not enabled for device runtime power management, not busy, not being used as wake up source and not have their state locked, depending on the power state.

I hadn't kept up to date with the changes to system PM, and I exclusively use runtime PM and PM_STATE_RUNTIME_IDLE, so I had an incorrect assumption that system PM was only managing drivers "inside" the SoC.

Commit:
doc: pm: document system pm limitations

Document the limitations of system power management when it comes to
transitions that are run from the idle thread.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

JordanYates commented:
The PR has been updated with a helper to determine if PM actions are allowed to block.
This gives actions an easy way to determine whether they can run the functions they need to for correct operation.
This pattern makes it explicit that system PM automatic transitions aren't supported on these types of devices.

gmarull reviewed (dismissed):
Changes look reasonable, I guess `pm_device_action_can_block` is the best we can do for now.

ceolin reviewed (changes requested):
That is not correct, pm_runtime can be used in `PRE_KERNEL`, which means that kernel services are not available yet, we can't simply check if it's called from idle thread.

ceolin commented:
> That is not correct, pm_runtime can be used in `PRE_KERNEL`, which means that kernel services are not available yet, we can't simply check if it's called from idle thread.

Also, this is a workaround for a bigger problem, is it addressing a real problem now ? We can have such thing if the problem is happening now and we have a use case, otherwise I don't thing we should rush a new API that is a workaround.

JordanYates commented:
> Also, this is a workaround for a bigger problem, is it addressing a real problem now ?

Yes, the problem in this PR. The power domain driver is not respecting timing parameters. This is required as many devices need to be totally depowered for a certain time before they will reset correctly. If the domain comes back up early the device can be in an undefined state.

> I don't thing we should rush a new API that is a workaround.

I agree that this is a workaround, but I don't see the core issue being resolved quickly, and the very existence of this function call makes it very obvious where updates need to be made when alternate options are available.

JordanYates commented:
> That is not correct, pm_runtime can be used in `PRE_KERNEL`, which means that kernel services are not available yet, we can't simply check if it's called from idle thread.

Updated to use the state information maintained by the PM subsystem instead.

gmarull commented on subsys/pm/device.c:
```diff
@@ -167,6 +169,11 @@ bool pm_device_is_any_busy(void)
 
+bool pm_device_action_can_block(void)
+{
+	return pm_state_next_get(_current_cpu->id)->state < PM_STATE_SUSPEND_TO_IDLE;
```
when system is resumed, the next state will always be reported as active, a device will never be called outside of system PM with the next state being `PM_STATE_RUNTIME_IDLE` because system will have resumed first (I think, system PM code is a bit messy and not easy to follow). 

some comments about this logic would help on understanding what's going on here

JordanYates replied:
The logic was intended to be that if the CPU is in ACTIVE or RUNTIME_IDLE mode it would be ok to block.
If this is not the case, then I need someone to explain exactly when these transitions can't block, because otherwise I'm just poking in the dark @ceolin

JordanYates replied:
> I am fine in having this function inside the pm subsystem, but this looks to me as a qutie generic kernel function.

I agree, this is a general query regarding the kernel state, so it belongs there

JordanYates commented on drivers/power_domain/power_domain_gpio.c:
```diff
@@ -22,28 +22,32 @@ struct pd_gpio_config {
 
+	/* Validate that blocking API's can be used */
+	if (!pm_device_action_can_block()) {
+		LOG_WRN("Blocking actions cannot run in this context");
+		return 0;
```
It depends what the failure mode should be. If you return an error my understanding is that the transition to a sleep mode will be aborted. With the current setup a problem is reported to the user and the domain is left on, but the board still goes to sleep.

gmarull replied:
-ENOTSUP is handled (device is ignored)

JordanYates replied:
Good point, done

gmarull commented:
LGTM, @ceolin please provide feedback on the `pm_device_action_can_block` current implementation.

ceolin reviewed (changes requested):
I have put some comments about the block API. The current pr does not work when the action is called before kernel fully initialized, which can happen with device runtime pm.

Commit:
kernel: add `k_can_yield` helper function

Implements a function that application and driver code can use to check
whether it is valid to yield (or block) in the current context. This
check is required for functions that can feasibly be run from multiple
contexts. The primary intended use case is power management transition
functions, which can be run by application code explicitly or
automatically in the idle thread by system PM.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
tests: kernel: validate `k_can_yield`

Validate the behaviour of `k_can_yield` in pre-kernel, ISR, and idle
thread and standard thread contexts.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

JordanYates commented:
The validity checker of whether we can block has been moved to `kernel.h` and given a suitably generic name `k_can_yield`. If this is likely to cause confusion with `k_yield`, we can use `k_can_block` or anything else that is suggested.

ceolin commented on include/kernel.h:
```diff
@@ -464,6 +464,19 @@ __syscall int32_t k_usleep(int32_t us);
+ *
+ * @return True if it is possible to yield in the current context, false otherwise.
+ */
+__syscall bool k_can_yield(void);
```
That is not something specific from this PR and other places are doing something similar I believe, but what is the point in having a syscall where there is no check and data crossing boundaries.

JordanYates replied:
Good point, I can't see any reason this needs to be a syscall, updated

ceolin reviewed (dismissed):
I don't know about the API name, maybe the core kernel folks have a better one. For me the pr looks ok.

JordanYates commented:
@andyross @dcpleung can I please get some feedback on the kernel portion

dcpleung commented:
I am okay with the naming, but @andyross can tell you if there is a better/alternative solution to this. Though I wonder if we can implement a wait function where it calls `k_sleep()` or `k_busy_wait()` depending on context (given that the wait is relatively short, say, maybe a couple ms).

JordanYates commented:
> Though I wonder if we can implement a wait function where it calls k_sleep() or k_busy_wait() depending on context (given that the wait is relatively short, say, maybe a couple ms).

While that may work in this particular situation, it won't for the more general problem where device transitions want to use comms interfaces that block (UART, SPI, I2C).

gmarull commented on kernel/sched.c:
```diff
@@ -1325,6 +1325,16 @@ static inline void z_vrfy_k_thread_deadline_set(k_tid_t tid, int deadline)
 
+bool k_can_yield(void)
+{
+	if (k_is_pre_kernel() ||
+	    k_is_in_isr() ||
+	    z_is_idle_thread_object(_current)) {
+		return false;
+	}
+	return true;
```
nit:
```
	return !(k_is_pre_kernel() || k_is_in_isr() ||
		 z_is_idle_thread_object(_current));
```

JordanYates replied:
done

gmarull reviewed (dismissed):
lgtm, one minor suggestion

Commit:
power_domain: gpio: update include paths

Update the Zephyr include paths to be compatible with removing
`CONFIG_LEGACY_INCLUDE_PATH` in the future.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

ceolin reviewed (approved):
approving it again
